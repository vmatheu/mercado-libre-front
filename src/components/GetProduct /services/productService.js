import axios from 'axios';
import { PRODUCT_SERVICE_ENDPOINT_ID } from '../../constants';

const transformProductSearch = (respuesta) => ({
    id: respuesta.item.id,
    title: respuesta.item.title,
    currency: respuesta.item.price.currency,
    amount: respuesta.item.price.amount,
    decimals: respuesta.item.price.decimals,
    picture: respuesta.item.picture,
    condition: respuesta.item.condition,
    freeShipping: respuesta.item.freeShipping,
    description: respuesta.item.description
})



export const findProductById = async (search) => {
  const productServiceEnpoint = PRODUCT_SERVICE_ENDPOINT_ID;
  const data = await axios.get(`${productServiceEnpoint}/${search}`)
    .then(response => ({
      item: transformProductSearch(response.data),
      categories: response.data.categories,
    }))
    .catch(error => {
      console.error(error)
      return []
    });
  return data;
};

export default { findProductById };
