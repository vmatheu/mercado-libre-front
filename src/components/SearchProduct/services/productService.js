import axios from 'axios';
import { PRODUCT_SERVICE_ENDPOINT } from '../constants';

const transformProductSearch = (products) => products.map(
  product => ({
    id: product.item.id,
    title: product.item.title,
    currency: product.item.price.currency,
    amount: product.item.price.amount,
    decimals: product.item.price.decimals,
    picture: product.item.picture,
    condition: product.item.condition,
    freeShipping: product.item.free_shipping
  })
);

export const findProductBySearch = async (search) => {
  const productServiceEnpoint = PRODUCT_SERVICE_ENDPOINT.development.url;
  const data = await axios.get(`${productServiceEnpoint}${search}`)
    .then(response => transformProductSearch(response.data))
    .catch(error => ([]));
  return data;
};

export default { findProductBySearch };
