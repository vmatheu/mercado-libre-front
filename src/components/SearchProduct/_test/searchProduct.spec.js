import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import productService from '../services/productService';
import { SearchProduct, onClick } from '..';

jest.mock('../services/productService');

describe('<SearchProduct/>', () => {
  describe('With search product by id', () => {
    it('should have the same ui structure', () => {
      const wrapper = shallow(<SearchProduct />);
      const tree = toJson(wrapper);
      expect(tree).toMatchSnapshot();
    });
  });

  describe('onClick action', () => {
    beforeEach(() => {
      productService.findProductBySearch = jest.fn();
    });

    it('should calling product service', async () => {
      await onClick('111', jest.fn());
      expect(productService.findProductBySearch).toHaveBeenCalled();
    });

    it('should calling product service when use ñ character', async () => {
      await onClick('aaañ', jest.fn());
      expect(productService.findProductBySearch).toHaveBeenCalled();
    });

    it('should not calling product service when search contains not alphanumeric characters', async () => {
      await onClick('21-', jest.fn());
      expect(productService.findProductBySearch).not.toHaveBeenCalled();
    });

    it('should calling product service with lower case character', async () => {
      await onClick('AAAA', jest.fn());
      expect(productService.findProductBySearch).toHaveBeenCalledWith('aaaa');
    });

    it('should not calling product service when search contains alphanumeric characters but lenght is less than three', async () => {
      await onClick('1a', jest.fn());
      expect(productService.findProductBySearch).not.toHaveBeenCalled();
    });

    it('should calling product service when search contains only numbers', async () => {
      await onClick('1', jest.fn());
      expect(productService.findProductBySearch).toHaveBeenCalled();
    });
  });
});
