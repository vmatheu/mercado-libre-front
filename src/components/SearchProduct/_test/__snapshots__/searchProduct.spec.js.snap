// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchProduct/> With search product by id should have the same ui structure 1`] = `
<div
  className="container"
>
  <Loading
    loading={false}
  />
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-md-5 "
    >
      <p>
        Ingresar busqueda de Productos
      </p>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-md-5"
    >
      <input
        className="form-control mcl-search-product-input-search"
        id="inputSearchId"
      />
      <input
        onClick={[Function]}
        text="Buscar"
        type="button"
      />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="mcl-search-product-bar-space"
    />
  </div>
  <div
    className="container mcl-search-product-productList"
  >
    <ProductList
      products={Array []}
    />
  </div>
</div>
`;
